lib/tsconfig-parser.js

```diff
@@ -0,0 +1,139 @@
+//------------------------------------------------------------------------------
+
+/**
+ * Maps tsconfig paths to their corresponding file contents and resulting watches
+ * @type {Map<string, ts.WatchOfConfigFile<ts.SemanticDiagnosticsBuilderProgram>>}
+ */
+const knownWatchProgramMap = new Map();
+
+/**
+ * Maps file paths to their set of corresponding watch callbacks
+ * There may be more than one per file if a file is shared between projects
+ * @type {Map<string, ts.FileWatcherCallback>}
+ */
+const watchCallbackTrackingMap = new Map();
+
+/**
+ * Holds information about the file currently being linted
+ * @type {{code: string, filePath: string}}
+ */
+const currentLintOperationState = {
+  code: '',
+  filePath: ''
+};
+
+/**
+ * Appropriately report issues found when reading a config file
+ * @param {ts.Diagnostic} diagnostic The diagnostic raised when creating a program
+ * @returns {void}
+ */
+function diagnosticReporter(diagnostic) {
+  throw new Error(
+    ts.flattenDiagnosticMessageText(diagnostic.messageText, ts.sys.newLine)
+  );
+}
+
+const noopFileWatcher = { close: () => {} };
+
+/**
+ * Calculate project environments using options provided by eslint and paths from config
+ * @param {string} code The code being linted
+ * @param {Object} options Options provided by ESLint core
+ * @param {string} options.cwd The current working directory for the eslint process
+ * @param {string} options.filePath The path of the file being parsed
+ * @param {Object} extra Validated parser options
+ * @param {string[]} extra.project Provided tsconfig paths
+ * @returns {ts.Program[]} The programs corresponding to the supplied tsconfig paths
+ */
+module.exports = function calculateProjectParserOptions(code, options, extra) {
+  const results = [];
+  const cwd = options.cwd || process.cwd();
+
+  // preserve reference to code and file being linted
+  currentLintOperationState.code = code;
+  currentLintOperationState.filePath = options.filePath;
+
+  // Update file version if necessary
+  // TODO: only update when necessary, currently marks as changed on every lint
+  const watchCallback = watchCallbackTrackingMap.get(options.filePath);
+  if (typeof watchCallback !== 'undefined') {
+    watchCallback(options.filePath, ts.FileWatcherEventKind.Changed);
+  }
+
+  for (let tsconfigPath of extra.project) {
+    // if absolute paths aren't provided, make relative to cwd
+    if (!path.isAbsolute(tsconfigPath)) {
+      tsconfigPath = path.join(cwd, tsconfigPath);
+    }
+
+    const existingWatch = knownWatchProgramMap.get(tsconfigPath);
+
+    if (typeof existingWatch !== 'undefined') {
+      // get new program (updated if necessary)
+      results.push(existingWatch.getProgram().getProgram());
+      continue;
+    }
+
+    // create compiler host
+    const watchCompilerHost = ts.createWatchCompilerHost(
+      tsconfigPath,
+      /*optionsToExtend*/ undefined,
+      ts.sys,
+      ts.createSemanticDiagnosticsBuilderProgram,
```

Comment:
I don't see you reporting semantic errors? Is that by design? If yes then you could use createAbstractBuilder which just updates program instead of caching semantic diagnostics etc.

Reply:
Related to https://github.com/uniqueiniquity/typescript-estree/pull/1#discussion_r226807552, I'm not necessarily reporting semantic errors, but I can foresee a lint rule that takes in a program and reports its diagnostics. So I'd like to leave that option open for users. I don't think it would make sense for a lint rule to cause emit to happen, so this seemed like the best `createProgram` to go with.

Reply:
See https://github.com/uniqueiniquity/typescript-estree/pull/1#discussion_r227055300